src/librustc/traits/structural_impls.rs

```diff
@@ -618,3 +618,186 @@ impl<'tcx> TypeFoldable<'tcx> for traits::ObligationCause<'tcx> {
         self.code.visit_with(visitor)
     }
 }
+
+impl<'tcx> fmt::Display for traits::WhereClauseAtom<'tcx> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use traits::WhereClauseAtom::*;
+        match *self {
+            Implemented(ref trait_ref) => write!(fmt, "Implemented({})", trait_ref),
+            ProjectionEq(ref projection) => write!(fmt, "ProjectionEq({})", projection),
+        }
+    }
+}
+
+impl<'tcx> fmt::Display for traits::DomainGoal<'tcx> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use traits::DomainGoal::*;
+        use traits::WhereClauseAtom::*;
+        match *self {
+            Holds(wc) => write!(fmt, "{}", wc),
+            WellFormed(Implemented(ref trait_ref)) => write!(fmt, "WellFormed({})", trait_ref),
+            WellFormed(ProjectionEq(ref projection)) => write!(fmt, "WellFormed({})", projection),
+            FromEnv(Implemented(ref trait_ref)) => write!(fmt, "FromEnv({})", trait_ref),
+            FromEnv(ProjectionEq(ref projection)) => write!(fmt, "FromEnv({})", projection),
+            WellFormedTy(ref ty) => write!(fmt, "WellFormed({})", ty),
+            FromEnvTy(ref ty) => write!(fmt, "FromEnv({})", ty),
+            RegionOutlives(ref predicate) => write!(fmt, "RegionOutlives({})", predicate),
+            TypeOutlives(ref predicate) => write!(fmt, "TypeOutlives({})", predicate),
+        }
+    }
+}
+
+impl fmt::Display for traits::QuantifierKind {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use traits::QuantifierKind::*;
+        match *self {
+            Universal => write!(fmt, "forall"),
+            Existential => write!(fmt, "exists"),
+        }
+    }
+}
+
+impl<'tcx> fmt::Display for traits::LeafGoal<'tcx> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use traits::LeafGoal::*;
+        match *self {
+            DomainGoal(ref domain_goal) => write!(fmt, "{}", domain_goal),
+        }
+    }
+}
+
+impl<'tcx> fmt::Display for traits::Goal<'tcx> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        use traits::Goal::*;
+        match *self {
+            Implies(ref hypotheses, ref goal) => {
+                write!(fmt, "if (")?;
+                for (index, hyp) in hypotheses.iter().enumerate() {
+                    if index > 0 {
+                        write!(fmt, ", ")?;
+                    }
+                    write!(fmt, "{}", hyp)?;
+                }
+                write!(fmt, ") {{ {} }}", goal)
+            }
+            And(ref goal1, ref goal2) => write!(fmt, "({}, {})", goal1, goal2),
```

Comment:
Nit: I've been moving towards adopting `&&` to mean "and" instead of the Prolog `,`... thoughts?